### ISSUE TITLE
Figure out how to mock API request in next.js

### ISSUE THREAD

**Issue by dadiorchen** [Greenstand/treetracker-web-map-client#157]
We need to figure out how to set up next.js, so we can test pages with mock API data. 
For the integration test in https://github.com/Greenstand/treetracker-wallet-web it's easy to write integration test against pages/paths, by using Cypress `intercept` API, cuz the API request is fired by client page, so we just simply intercept the API and return mock data.
But for next.js, we render some part of page on the server side(SSR), then it's impossible for Cypress to intercept the http request (because it happened in next server) 
We need to figure out a way to easily mock the API in the integration test.

---
Some hints:
* Please read our [readme](https://github.com/Greenstand/treetracker-web-map-client#treetracker-web) for more information/guide/tutorial.
* Here is [an engineering book](https://greenstand.gitbook.io/engineering/) in Greenstand.
* To know more about our organization, visit our [website](https://greenstand.org).

**Comment by dadiorchen** [comment]
Some clues:
This https://glebbahmutov.com/blog/mock-network-from-server/ is talking about move the next process into Cypress process (as subprocess), this is a good idea, but this only problem I think is that: can we easily write test by automatically update the next process, I mean, when we run the test and making changes in code, say, change some code for the SSR, then can we make the test re-run and get the result immediately.
You know, currently, we run Cypress test and the next server separately ( npm run cypress:open, npm run dev, the second is for starting next) in this way, the next process is automatically refreshed (hot reload), this is convenient as we run the server and coding and making changes.

**Comment by dadiorchen** [comment]
Or we can consider leverage some things from nextjs API  https://nextjs.org/docs/api-routes/api-middlewares

**Comment by QuaidBartolomei** [comment]
I will work on this.

**Comment by dadiorchen** [comment]
Yes, this issue is about finding a good way to mock the data in tests, so we don't rely on external service including the `mock-server`. For client side render, it's easy to do it through `cy.intercept` but it's a bit hard to do it on the case of SSR, do you have any idea?

**Comment by QuaidBartolomei** [comment]
Ok I understand now. I will work on a solution based on the blog post you shared.

**Comment by QuaidBartolomei** [comment]
@dadiorchen I tested the example in the blog post you shared. This method will auto-reload the tests when developers make any frontend changes (i.e. changes to a component). If changes to the Nextjs api (i.e. changes to a getServerSideProps function) are made then the cypress tests must be manually rerun with the refresh button inside the cypress browser window. This doesn't seem so bad to me, should I go ahead and implement it?

**Comment by QuaidBartolomei** [comment]
I implemented code-watching and next api mocking. Now integration tests (cypress open) can be run without `npm run dev` and without `npm run mock-server`. Any changes to files in the 'src' directory while cypress is running will restart the tests.
